docs/Testing.md

```diff
@@ -0,0 +1,454 @@
+not well-suited to run on iOS directly.
+
+To handle this, Doppl provides a test runner and means of getting at a mock
+`Application` object. When running the tests on Android, the mock comes
+from Robolectric. When running the tests on iOS, the mock comes from Doppl itself.
+
+### Using the Test Runner
+
+First, add Robolectric to your `dependencies`
+(e.g., `testImplementation  "org.robolectric:robolectric:3.3.2"`), if you do not
+have it for other reasons already.
+
+Then, create a JUnit test class that uses 
+`DopplContextDelegateTestRunner` as the test runner:
+
+```java
+@RunWith(DopplContextDelegateTestRunner.class)
+public class ExampleContextTest {
+  @Test
+  public void testSomething() {
+    // TODO: a really cool test
+  }
+}
+```
+
+Here, `DopplContextDelegateTestRunner` comes from the `co.doppl.lib:androidbasetest`
+dependencies that you should have set up for the `testImplementation` and
+the `testDoppl` dependency scopes.
+
+### Getting an Application Context
+
+Then, your test methods can use `DopplRuntimeEnvironment.getApplication()`
+to get a mock `Application` singleton:
+
+```java
+@RunWith(DopplContextDelegateTestRunner.class)
+public class ExampleContextTest {
+  @Test
+  public void testSomething() {
+    final SharedPreferences prefs=
+      DopplRuntimeEnvironment
+        .getApplication()
+        .getSharedPrefences(TEST_PREFS, Context.MODE_PRIVATE);
+
+    // TODO: a really cool test involving SharedPreferences
+  }
+}
+```
+
+Doppl will arrange to give you a suitable mock, both on Android (via
+Robolectric) or on iOS.
+
+## Using Spies with Mockito
+
+Doppl integrates with J2objc's support for Mockito. However, at the present time,
+that support is limited to mocks. If you want to spy an object &mdash; perhaps
+to confirm that methods were invoked &mdash; J2objc does not support that.
+
+However, *Doppl* supports object spies, though it takes a bit more work to set up,
+beyond simply having Mockito as a test dependency (which is all that you need
+for simple mocks). This is through a `@MockGen` annotation, backed by a compile-time
+annotation processor that generates code to support object spying on iOS.
+
+Note that this is a stop-gap implementation, while improvements to J2objc's
+Mockito support are still in progress.
+
+### Writing a Test Using a Spy
+
+On the Android side, you can go ahead and create your unit test that
+uses Mockito's spy capability:
+
+```java
+@Test
+public void timer() throws InterruptedException {
+  ScheduledExecutorService scheduler=Executors.newSingleThreadScheduledExecutor();
+  Runnable thingy=spy(new Runnable() {
+    @Override
+    public void run() {
+      // TODO: something useful
+    }
+  });
+
+  verify(thingy, never()).run();
+  scheduler.schedule(thingy, 200, TimeUnit.MILLISECONDS);
+  Thread.sleep(500);
+  verify(thingy, times(1)).run();
+}
+```
+
+This should run fine in your Android unit tests... on Android. It will not
+work on iOS, until we add in `@MockGen`.
```

Comment:
Beats me. I'm just the scribe. :-)

Reply:
We're using the Mockito provided by J2objc out of the box. It doesn't mock objects in the same what a regular JVM and the way an Android mock system would work. You can provide a plugin implementation to do that (see here https://github.com/linkedin/dexmaker). We would need to create one. Spy will sometimes run, but probably won't have correct results, and sometimes crashes due to zombie references. It might run, but no, it doesn't work.